hudi-common/src/main/java/org/apache/hudi/common/table/read/FileGroupRecordBuffer.java

```diff
@@ -331,6 +343,64 @@ protected Option<BufferedRecord<T>> doProcessNextDataRecord(BufferedRecord<T> ne
     }
   }
 
+  /**
+   * Merge records based on partial update mode.
+   * Note that {@param newRecord} refers to the record with higher commit time
+   * if COMMIT_TIME_ORDERING mode is used, or higher event time if EVENT_TIME_ORDERING mode us used.
+   */
+  private void updatePartiallyIfNeeded(BufferedRecord<T> newRecord,
+                                       BufferedRecord<T> oldRecord,
+                                       Schema newSchema,
+                                       Schema oldSchema,
+                                       PartialUpdateMode partialUpdateMode) {
+    if (newRecord.isDelete()) {
+      return;
+    }
+
+    List<Schema.Field> fields = newSchema.getFields();
+    switch (partialUpdateMode) {
+      case NONE:
+      case KEEP_VALUES:
+      case FILL_DEFAULTS:
+      case COLUMN_FAMILY:
+        // No-op for these modes
+        break;
+
+      case IGNORE_DEFAULTS:
+        for (Schema.Field field : fields) {
+          String fieldName = field.name();
+          Object defaultValue = field.defaultVal();
+          Object newValue = readerContext.getValue(
+              newRecord.getRecord(), newSchema, fieldName);
+          if (defaultValue == newValue) {
+            Object oldValue = readerContext.getValue(oldRecord.getRecord(), oldSchema, fieldName);
+            readerContext.setValue(
+                newRecord.getRecord(), newSchema, fieldName, oldValue);
+          }
+        }
+        break;
+
+      case IGNORE_MARKERS:
+        if (partialUpdateCustomMarkerValue == null) {
+          throw new IllegalStateException(
+              "For 'IGNORE_MARKERS' mode, custom marker value must be defined");
+        }
+        for (Schema.Field field : fields) {
+          String fieldName = field.name();
+          Object newValue = readerContext.getValue(newRecord.getRecord(), newSchema, fieldName);
+          if (partialUpdateCustomMarkerValue.equals(readerContext.castToString(newValue))) {
```

Comment:
we should only do this for `String` and `Bytes` datatype. check out https://github.com/apache/hudi/blob/33a7b845bcd30f823d026b9f19e20554fe3cf10e/hudi-common/src/main/java/org/apache/hudi/common/model/debezium/PostgresDebeziumAvroPayload.java#L140